### PR TITLE
Update audit scope based on audit-v1.0.0-rc.0 tag

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR updates the audit scope to use the [`audit-v1.0.0-rc.0` branch](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0).

This removes a significant amount of duplicate cryptographic code, and fixes multiple production consensus rule and security issues. It also includes some changes required to get other changes to merge, build, or pass lints.

You can see a full list of changes on top of the `v1.0.0-rc.0` release here:
- https://github.com/ZcashFoundation/zebra/pull/6108/commits

## Solution

- Update Zebra versions and links to the audit branch
- Add the new `reddsa` ZF dependency
- Add a list of significant changes that couldn't be merged into the `audit-v1.0.0-rc.0` branch
- Note that duplicate consensus-critical dependencies will be fixed soon
- Note that tests might not build, run, or pass

## Review

We need to decide if these changes are in scope:
- [ ] ZF dependency `reddsa` 0.4.0, which replaces some removed `zebra-chain` code
- [x] the changes that couldn't be merged into the `audit-v1.0.0-rc.0` branch due to conflicts

### Reviewer Checklist

  - [ ] Are the PR labels correct?
  - [ ] Do the audit changes make sense?

### CODE REVIEW

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 1]
```diff
@@ -11,22 +11,23 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
+| zebra-script | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-script/src) |
+| zebra-state | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-state/src) |
+| zebrad | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebrad/src) |
 
 ### Zcash/ZF dependencies
 
 | Name | Version | Notes
 |------| ------- | -----
 | ed25519-zebra | [3.1.0](https://github.com/ZcashFoundation/ed25519-zebra/tree/3.1.0/src)
+| reddsa | [0.4.0](https://github.com/ZcashFoundation/reddsa/tree/0.4.0/src)
```
Comment (teor2345) [inline book/src/dev/zebra-dependencies-for-audit.md]: Do we want to audit `reddsa` with Zebra now, or as part of a future Zebra or FROST audit?

If we don't want to audit it now, we need to move it to the "Out of Scope" section.

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 2]
```diff
@@ -44,14 +45,24 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
 | zcash_script | 0.1.8 || <i>The C++ parts of `zcashd` got audited as part of the ECC audit, so we only need to audit: <br />- [zcash_script.cpp](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.cpp), <br />- [zcash_script.h](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.h), and <br />- [the rust code in the zcash_script crate](https://github.com/ZcashFoundation/zcash_script/tree/v0.1.8/src).</i>
 
+Note: there are duplicate `zcash_primitives` and `zcash_proofs` dependencies in Zebra's audit and development branches, [this will get fixed](https://github.com/ZcashFoundation/zebra/issues/6107) after the `zcashd` 5.4.0 release.
+
+---
+## Not Included
+
```
Comment (teor2345) [inline book/src/dev/zebra-dependencies-for-audit.md:53]: I don't think we want to audit these PRs, because they include changes that are out of scope. But we probably want to keep a record of them somewhere, so we know where to start checking for existing fixes to audit bugs.

```suggestion

The changes in these PRs are out of scope for the audit. When the Zebra team checks for bugs that have already been fixed, we can check these PRs, and any changes after commit [c4032e2b](https://github.com/ZcashFoundation/zebra/commit/c4032e2b7f6dbee8a9480d3c978c70a3cfc3332c).

```

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 This is a list of production Rust code that is in scope and out of scope for Zebra's first audit.
 
-Test code, deployment configurations, and other configuration files in the `zebra` repository are out of scope.
+Test code, deployment configurations, and other configuration files in the `zebra` repository are out of scope. Due to the way we've created the `audit-v1.0.0-rc.0` branch, they might not compile, run, or pass.
```
Comment (arya2) [inline book/src/dev/zebra-dependencies-for-audit.md]: Nitpick:

```suggestion
Test code, deployment configurations, and other configuration files in the `zebra` repository are out of scope. Due to the way we've created the `audit-v1.0.0-rc.0` branch, tests might not compile, run, or pass.
```

If it would be useful to auditors and it's easy to fix, we may want to update the branch so they'll compile.

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 4]
```diff
@@ -11,22 +11,23 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
+| tower-batch | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/tower-batch/src) |
+| tower-fallback | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/tower-fallback/src) | 
+| zebra-chain | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-chain/src) |
+| zebra-consensus | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-consensus/src) |
+| zebra-network | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-network/src) |
+| zebra-node-services | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-node-services/src) |
+| zebra-rpc | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-rpc/src) |
+| zebra-script | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-script/src) |
+| zebra-state | [audit-v1.0.0-rc.0](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-state/src) |
```
Comment (arya2) [inline book/src/dev/zebra-dependencies-for-audit.md:22]: Should these match up with the crate version? If not, should the column headers be changed to "branch" instead of "version"?

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 1 above)
Comment (arya2) [reply book/src/dev/zebra-dependencies-for-audit.md]: We should discuss this in our sync meeting.

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 2 above)
Comment (arya2) [reply book/src/dev/zebra-dependencies-for-audit.md:53]: This mention should come in handy for auditors.

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 3 above)
Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md]: > If it would be useful to auditors and it's easy to fix, we may want to update the branch so they'll compile.

The priority for this branch was significant changes in auditable code.

So to avoid merge conflicts in the tests, I didn't apply any test changes or fixes, unless they were part of a significant change PR. So even if we get the tests compiling, I'm not sure if they will pass.

We also agreed to limit the time we spent on the audit branch to 1 day, so we only have a few hours left to make any changes.

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 1 above)
Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md]: > We should discuss this in our sync meeting.

It's first on the agenda, because we have to decide by tomorrow.

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 5]
```diff
@@ -44,14 +45,24 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
 | zcash_script | 0.1.8 || <i>The C++ parts of `zcashd` got audited as part of the ECC audit, so we only need to audit: <br />- [zcash_script.cpp](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.cpp), <br />- [zcash_script.h](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.h), and <br />- [the rust code in the zcash_script crate](https://github.com/ZcashFoundation/zcash_script/tree/v0.1.8/src).</i>
 
+Note: there are duplicate `zcash_primitives` and `zcash_proofs` dependencies in Zebra's audit and development branches, [this will get fixed](https://github.com/ZcashFoundation/zebra/issues/6107) after the `zcashd` 5.4.0 release.
```
Comment (teor2345) [inline book/src/dev/zebra-dependencies-for-audit.md]:
```suggestion
Note: there are duplicate `zcash_primitives`, `zcash_proofs`, and `reddsa` dependencies in Zebra's audit and development branches, [this will get fixed](https://github.com/ZcashFoundation/zebra/issues/6107) after the `zcashd` 5.4.0 release.
```

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 4 above)
Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md:22]: This depends on if we want to freeze the branch as a tag - I'll ask this at the meeting as well.

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 1 above)
Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md]: We said `reddsa` and `redjubjub` would be nice to have, Deirdre is going to add the exact scope to this PR.

File: book/src/dev/zebra-dependencies-for-audit.md (hunk 4 above)
Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md:22]: We're going to replace the branch with a tag once it's been reviewed and is final. So it will appear as a version on GitHub.

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 6]
```diff
@@ -35,7 +36,7 @@ Test code, deployment configurations, and other configuration files in the `zebr
 
 | Name | Version | Notes
 |------| ------- | -----
-| zebra-utils | 1.0.0-beta.17 | <i>Only the [zebra-checkpoints](https://github.com/ZcashFoundation/zebra/tree/v1.0.0-rc.1/zebra-utils/src/bin/zebra-checkpoints) utility needs to be audited.</i>
+| zebra-utils | audit-v1.0.0-rc.0 | <i>Only the [zebra-checkpoints](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-utils/src/bin/zebra-checkpoints) utility needs to be audited.</i>
```
Comment (dconnolly) [inline book/src/dev/zebra-dependencies-for-audit.md:39]:
```suggestion
| zebra-utils | audit-v1.0.0-rc.0 | <i>Only the [zebra-checkpoints](https://github.com/ZcashFoundation/zebra/tree/audit-v1.0.0-rc.0/zebra-utils/src/bin/zebra-checkpoints) utility needs to be audited.</i>
| redjubjub | 0.5.0 | <i>All files should be audited EXCEPT the [FROST code](- https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/frost.rs), the [messages](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages*), and the [signing code]( https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/signing.rs)</i>
| reddsa | 0.4.0 | <i>All files should be audited EXCEPT the [signing code]( https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/signing.rs), and the [Sapling code]( https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/sapling.rs)</i>
```

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 7]
```diff
@@ -35,14 +35,27 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
 | zcash_script | 0.1.8 || <i>The C++ parts of `zcashd` got audited as part of the ECC audit, so we only need to audit: <br />- [zcash_script.cpp](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.cpp), <br />- [zcash_script.h](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.h), and <br />- [the rust code in the zcash_script crate](https://github.com/ZcashFoundation/zcash_script/tree/v0.1.8/src).</i>
+| reddsa | [0.4.0](https://github.com/ZcashFoundation/reddsa/tree/0.4.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>This code was moved from `zebra/zebra-chain/src/primitives/redpallas` into a separate crate after the Zebra `v1.0.0-rc.0` release. A previous version of this code was audited as the `redjubjub` crate.<br />The FROST messages in [messages.rs](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/messages.rs) are out of scope.</i>
```
Comment (dconnolly) [inline book/src/dev/zebra-dependencies-for-audit.md]:
```suggestion
| redjubjub | [0.5.0](https://github.com/ZcashFoundation/redjubjub/tree/0.5.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>All files should be audited EXCEPT:<br />-the [signing code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/signing_key.rs)<br />- the [FROST code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/frost.rs), and<br />- the FROST messages [module](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages.rs) and [directory](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages)</i>
| reddsa | [0.4.0](https://github.com/ZcashFoundation/reddsa/tree/0.4.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>This code was moved from `zebra/zebra-chain/src/primitives/redpallas` into a separate crate after the Zebra `v1.0.0-rc.0` release. A previous version of this code was audited as the `redjubjub` crate.<br />All files should be audited EXCEPT:<br />- the [signing code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/signing_key.rs), and<br />- the [Sapling code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/sapling.rs)</i>
```

Comment (teor2345) [reply book/src/dev/zebra-dependencies-for-audit.md]: I just did a quick edit for formatting and fixing links, let's see how it goes in markdown.

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 8]
```diff
@@ -31,18 +31,32 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
+| reddsa | [0.4.0](https://github.com/ZcashFoundation/reddsa/tree/0.4.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>This code was moved from `zebra/zebra-chain/src/primitives/redpallas` into a separate crate after the Zebra `v1.0.0-rc.0` release. A previous version of this code was audited as the `redjubjub` crate.<br />All files should be audited EXCEPT:<br />- the [signing code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/signing.rs), and<br />- the [Sapling code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/sapling.rs)</i>
```
Comment (dconnolly) [inline book/src/dev/zebra-dependencies-for-audit.md]:
```suggestion
| redjubjub | [0.5.0](https://github.com/ZcashFoundation/redjubjub/tree/0.5.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>All files should be audited EXCEPT:<br />- the [signing code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/signing_key.rs)<br />- the [FROST code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/frost.rs), and<br />- the FROST messages [module](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages.rs) and [directory](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages)</i>
| reddsa | [0.4.0](https://github.com/ZcashFoundation/reddsa/tree/0.4.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) | <i>This code was moved from `zebra/zebra-chain/src/primitives/redpallas` into a separate crate after the Zebra `v1.0.0-rc.0` release. A previous version of this code was audited as the `redjubjub` crate.<br />All files should be audited EXCEPT:<br />- the [signing code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/signing_key.rs), and<br />- the [Sapling code](https://github.com/ZcashFoundation/reddsa/blob/0.4.0/src/sapling.rs)</i>
```

File: book/src/dev/zebra-dependencies-for-audit.md [hunk 9]
```diff
@@ -31,18 +31,32 @@ Test code, deployment configurations, and other configuration files in the `zebr
...
 | zcash_script | 0.1.8 || <i>The C++ parts of `zcashd` got audited as part of the ECC audit, so we only need to audit: <br />- [zcash_script.cpp](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.cpp), <br />- [zcash_script.h](https://github.com/ZcashFoundation/zcash_script/blob/v0.1.8/depend/zcash/src/script/zcash_script.h), and <br />- [the rust code in the zcash_script crate](https://github.com/ZcashFoundation/zcash_script/tree/v0.1.8/src).</i>
+| redjubjub | [0.5.0](https://github.com/ZcashFoundation/redjubjub/tree/0.5.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) <i>(FROST only)</i> | <i>All files should be audited EXCEPT:<br />-the [signing code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/signing_key.rs)<br /> - the [FROST code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/frost.rs), and<br />- the FROST messages [module](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages.rs) and [directory](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages)</i>
```
Comment (arya2) [inline book/src/dev/zebra-dependencies-for-audit.md]: Very minor nitpick:
```suggestion
| redjubjub | [0.5.0](https://github.com/ZcashFoundation/redjubjub/tree/0.5.0/src) | [jp](https://github.com/ZcashFoundation/redjubjub/raw/main/zcash-frost-audit-report-20210323.pdf) <i>(FROST only)</i> | <i>All files should be audited EXCEPT:<br />- the [signing code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/signing_key.rs)<br /> - the [FROST code](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/frost.rs), and<br />- the FROST messages [module](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages.rs) and [directory](https://github.com/ZcashFoundation/redjubjub/blob/0.5.0/src/messages)</i>
```

**Review by arya2 — DISMISSED** [review]

**Review by arya2 — APPROVED** [review]

**Review by dconnolly — COMMENTED** [review]
👍
